resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -256,8 +321,153 @@ private[spark] class KubernetesClusterSchedulerBackend(
+            case _ => false
+          }
+        }
+
+        override def apply(msg: Any): Unit = {
+          msg match {
+            case RetrieveSparkAppConfig(executorId) =>
+              EXECUTOR_MODIFICATION_LOCK.synchronized {
+                var resolvedProperties = sparkProperties
+                val runningExecutorPod = kubernetesClient
+                  .pods()
+                  .withName(runningExecutorPods(executorId).getMetadata.getName)
+                  .get()
+                val nodeName = runningExecutorPod.getSpec.getNodeName
+                val shuffleServicePods = kubernetesClient
+                  .pods()
+                  .inNamespace(shuffleServiceConfig.get.shuffleNamespace)
+                  .withLabels(shuffleServiceConfig.get.shuffleLabels.asJava)
+                  .list()
+
+                // Filter only those pods which are posting Ready status and on the same
+                // node as our executor.
+                val readyPods = shuffleServicePods
+                  .getItems.asScala.toList.filter(pod => Readiness.isPodReady(pod))
+                val colocatedPods = readyPods.filter(_.getSpec.getNodeName == nodeName)
+
+                colocatedPods match {
+                  case Nil =>
+                    logError(s"Cannot find co-located shuffle pod on node ${nodeName}")
+                    throw new SparkException(s"Cannot find co-located shuffle pod on node " +
+                      s"${nodeName}")
+
+                  case shufflePod :: Nil =>
+                    val shuffleIP = conf
+                      .get(KUBERNETES_SHUFFLE_SVC_IP)
+                      .getOrElse(shufflePod.getStatus.getPodIP)
+
+                    resolvedProperties = resolvedProperties ++ Seq(
+                      (SPARK_SHUFFLE_SERVICE_HOST.key, shuffleIP))
+
+                  case _ =>
+                    val podsList = colocatedPods.map(_.getStatus.getPodIP).mkString(",")
+                    logError(s"Ambiguous specification of shuffle service pod. " +
+                      s"Found multiple matching pods: ${podsList}")
+                    throw new SparkException(s"Ambiguous specification of shuffle service pod. " +
+                      s"Found multiple matching pods: ${podsList}")
+
+                }
+
+                val reply = SparkAppConfig(resolvedProperties,
+                  SparkEnv.get.securityManager.getIOEncryptionKey())
+                context.reply(reply)
+              }
+          }
+        }
+      }.orElse(super.receiveAndReply(context))
+    }
+  }
+
+  /*
+ * KubernetesAllocator class watches all the executor pods associated with
+ * this SparkJob and creates new executors when it is appropriate.
+ */
+  private[spark] class KubernetesAllocator(
+    client: KubernetesClient,
+    interval: Int,
+    backend: KubernetesClusterSchedulerBackend) extends Logging {
+
+    private val scheduler = Executors.newScheduledThreadPool(1)
```

Comment:
Wondering, if this should be a daemon thread. If yes, then use ThreadUtils from Spark packages to create a "named" daemon thread.